optuna/_convert_positional_args.py

```diff
@@ -37,22 +61,66 @@ def convert_positional_args(
     *,
     previous_positional_arg_names: Sequence[str],
     warning_stacklevel: int = 2,
+    deprecated_version: str | None = None,
+    removed_version: str | None = None,
 ) -> "Callable[[Callable[_P, _T]], Callable[_P, _T]]":
     """Convert positional arguments to keyword arguments.
 
     Args:
-        previous_positional_arg_names: List of names previously given as positional arguments.
-        warning_stacklevel: Level of the stack trace where decorated function locates.
+        previous_positional_arg_names:
+            List of names previously given as positional arguments.
+        warning_stacklevel:
+            Level of the stack trace where decorated function locates.
+        deprecated_version:
+            Version number in which the use of positional arguments is deprecated.
+        removed_version:
+            Version number in which the use of positional arguments was completely removed.
     """
 
+    if deprecated_version is None and removed_version is None:
+        pass
+    else:
+        if deprecated_version is None:
+            raise ValueError(
+                "deprecated_version must not be None when removed_version is specified"
+            )
+        if removed_version is None:
+            raise ValueError(
+                "removed_version must not be None when deprecated_version is specified"
+            )
+
+        _validate_version(deprecated_version)
+        _validate_version(removed_version)
+        _validate_two_version(deprecated_version, removed_version)
+
     def converter_decorator(func: "Callable[_P, _T]") -> "Callable[_P, _T]":
+        if deprecated_version or removed_version:
+            if func.__doc__ is None:
+                func.__doc__ = ""
+
+            note = _DEPRECATION_NOTE_TEMPLATE.format(
+                d_ver=deprecated_version, r_ver=removed_version
+            )
+            indent = _get_docstring_indent(func.__doc__)
+            func.__doc__ = func.__doc__.strip() + textwrap.indent(note, indent) + indent
+
         assert set(previous_positional_arg_names).issubset(set(signature(func).parameters)), (
             f"{set(previous_positional_arg_names)} is not a subset of"
             f" {set(signature(func).parameters)}"
         )
 
         @wraps(func)
         def converter_wrapper(*args: Any, **kwargs: Any) -> "_T":
+            if deprecated_version or removed_version:
+                warnings.warn(
+                    _DEPRECATION_WARNING_TEMPLATE.format(
+                        name=func.__name__,
+                        d_ver=deprecated_version,
+                        r_ver=removed_version,
+                    ),
+                    FutureWarning,
+                    stacklevel=warning_stacklevel,
+                )
```

Comment:
With this implementation, the warning will always be shown, so please change it so that it only appears for positional arguments.